README
Versión de fslurp para inversores fronius de la terraza.
llamada online
/home/u4477/fslurp/fslurp -b 19200 -r all -d , -p /dev/ttyS1

ver --> sendtoemoncms.sh
(lee los datos, los almacena en ficheros /home/u4477/fslurp/data)

para arrancarlo se hace con un cron. 

sudo crontab -e

* * * * * /home/u4477/fslurp/fslurp2emoncms.sh
*/5 * * * * /home/u4477/fslurp/fslurp2dataciemat.sh

la tarea fslurp2dataciemat creo ficheros de datos cada 5 minutos, con formato de fecha leible por humano

$ cd /wherever/you/put/it/fslurp-0.5
$ make clobber
$ make

And then to run it, from the same directory:

$ ./fslurp -h

This will display the following help message:

Usage: ./fslurp [-b baud][-d delimiter][-H][-h][-p device][-r report][-v]
-b baud Set the baud rate. Supported rates are:
2400, 4800, 9600, and 19200. Defaults to 9600.
-d delimter Set delimiter for CSV-style output. The
default is no delimiter, human-readable.
-H Display the delimited output header for the
current reporting mode. The -d option must
also be set.
-h Display this help message.
-p device Set the serial port device. Defaults to /dev/ttyS0.
-r report Set report type. Supported types are:
now, day, total, and all Defaults to now.
-v Enable verbose, debugging output.

Return codes:
0 Success.
1 Communications failure.
2 No active inverter found.
3 Usage error.
4 Fronius protocol error.

#=====================================================================================
del github original fslurp
#=====================================================================================
$RCSfile: README,v $
$Revision: 1.3 $
$Date: 2013/11/17 00:16:12 $

Introduction
============
fslurp is a program to pull the data from a Fronius inverter systems
via a serial port connection.

It supports both single and multiple inverter systems through the
Interface Card Easy and the Fronius Interface Card/DataLogger Box. Fronius
Sensor Card devices are also supported.

It supports both human-readable and delimited (CSV-style) text output.

Sample output can be seen by building project and exercising one
of the test simulator targets (e.g. simIFCEasy).

It is written in C++ and has been compiled on Linux.

Building
========

The GNU C++ compiler and the GNU make program
are required for building fslurp.

Un-tar the source.
cd to the top level directory extracted.
Type make to build.

Installing and running
======================

The fslurp binary requires no external files and can be run
from anywhere.  Copy it to wherever as appropriate.

Testing
=======

In addition to fslurp, the makefile builds several simulator versions
of fslurp (e.g. simIFCEasy). The simulators are identical to fslurp,
except that they each replace the serial communications with simulated
canned responses for a particular Fronius installation configuration.

The test target in the makefile exercises fslurp via these simulators.

David Stone
ds.fslurp@asif.com


fslurp_fv42/History
$RCSfile: History,v $
$Revision: 1.9 $
$Date: 2013/12/21 14:57:37 $

Version 2.1.3	131116	Added support/testing for RS-422 connected inverters.
			Fixed problem in fslurp2pvoutput script for
			fractional KWH.

Version 2.1.3	131116	Added 0x97 Sunpower SPR-3501F-2 to device table.

Version 2.1.2	121027	Makefile updates.
			Added README file.

Version 2.1.1	121019	Fixed problem with sensor card reporting when
			no sensor cards are present.
			Updated -H handling to also display the header
			when no devices are present.
			Testing updates.
			Man page updates.
			Makefile updates.

Version 2.1.0	121015	Added support for year data and report handling.
			Added year support to inverter/sensor simulation.

Version	2.0.9	121013	Implemented testing, updated make test target.
			Added sensor card data and report handling.
			Added sensor card command simulation.
			Format handling changes for widths of different report
			types in non-delimited output.

Version 2.0.8	121012	Reorganized to update formatting for the all report.
			Added initial version of the man page
			(thanks to Sergio Zanchetta <primes2h@gmail.com>).

Version 2.0.7	121007	Added support for -n deviceNumber.
			Simplified system reporting implementation by
			using DeviceIterator.
			Added InverterSingleTargetIterator and
			SensorCardSingleTargetIterator.
			Move exit code defines into an Args::ExitCodes enum.

Version 2.0.6	121006	Added support for -D deviceType option.
			Implemented delimited report headers for system and
			inverter reports.
			Abtracted Inverter and SensorCard report handling.
			Added DeviceIterator.
			Added a second inverter in the SolarNet simulator.


args.cpp
//
// fslurp, Copyright (c) 2012 David Stone <ds.fslurp@asif.com>
//
// $RCSfile: args.cpp,v $
// $Revision: 1.7 $
// $Date: 2012/10/15 15:44:37 $
//
// Permission to use, copy, modify, distribute, and sell this software and its
// documentation for any purpose is hereby granted without fee, provided that
// the above copyright notice appear in all copies and that both the
// copyright notice and this permission notice appear in supporting
// documentation.  No representations are made about the suitability of this
// software for any purpose.  It is provided "as is" without express or 
// implied warranty.
//

#include "fslurp.h"

const char* Args::programName_;

Args::Args(int argc, char* argv[])
{
	// set defaults

	serialPortName_ = "/dev/ttyS0";
	baudRate_ = "9600";
	reportTypeName_ = "system";
	reportType_ = ReportTypeSystem;
	deviceType_ = DeviceTypeInverter;
	deviceTargeting_ = AllInstances;
	timeFormat_ = "%c";
	delimiter_ = NULL;
	showReportHeader_ = false;
	verbose_ = false;

	parseArgs(argc, argv);

	// initialize global timestamp

	timestamp_ = time(NULL);
}

const char*
Args::getSerialPortName(void) const
{
	return serialPortName_;
}

const char*
Args::getDelimiter(void) const
{
	return delimiter_;
}

const char*
Args::getTimeFormat(void) const
{
	return timeFormat_;
}

time_t
Args::getTimestamp(void) const
{
	return timestamp_;
}

bool
Args::showReportHeader(void) const
{
	return showReportHeader_;
}

void
Args::usage(void)
{
	fprintf(stderr, 

	"Usage: %s\n"
	"       [-b baud] [-D device_type] [-d delimiter] [-H] [-h]\n"
	"       [-n device_number] [-p serial_port_device] [-r report]\n"
	"       [-t format] [-v]\n"
	"\n"
	"-b baud              Set the baud rate.  Supported rates are:\n"
	"                     2400, 4800, 9600, and 19200. Defaults to %s.\n"
	"-D device_type       inverter or sensor. Defaults to inverter.\n"
	"-d delimiter         Set delimiter for CSV-style output. The\n"
	"                     default is no delimiter, human-readable.\n"
	"-H                   Include the delimited output header for the\n"
	"                     current reporting mode.  The -d option must\n"
	"                     also be set.\n"
	"-h                   Display this help message.\n"
	"-n device_number     Run the report for an individual inverter or\n"
	"                     sensor card. The device number is specified in\n"
	"                     decimal, and can be seen in the\n"
	"                     The default is to report on all\n"
	"                     devices of the current type (see the -D option)\n"
	"-p serial_device     Set the serial port device. Defaults to %s.\n"
	"-r report            Set report type. Supported types are:\n"
	"                     system, now, day, year, total, and all.\n"
	"                     Defaults to %s. Year is not supported for\n"
	"                     Interface Card Easy systems.\n"
	"-t format            Set the format for the timestamp field,\n"
	"                     using the Linux strftime(3) format string.\n"
	"                     The default is \"%%c\", which looks like:\n"
	"                             Wed May 26 10:43:25 2010\n"
	"-v                   Enable verbose, debugging output.\n"
	"\n"
	"Return codes:\n"
	"\n"
	"%d                   Success.\n"
	"%d                   Communications failure.\n"
	"%d                   Inverter not found.\n"
	"%d                   Sensor card not found.\n"
	"%d                   Usage error.\n"
	"%d                   Fronius protocol error.\n",

	programName_,
	baudRate_,
	serialPortName_,
	reportTypeName_,
	ExitNormal,
	ExitCommFailure,
	ExitInverterNotFound,
	ExitSensorCardNotFound,
	ExitUsageError,
	ExitProtocolError);
}

void
Args::parseArgs(int argc, char* argv[])
{
	int	c;

	CHECK_PTR(argv);
	assert(argc >= 1);

	programName_ = argv[0];

	while (1)
	{
		c = getopt(argc, argv, "b:D:d:Hhn:p:r:t:v");

		if (c == -1)
		{
			break;
		}

		switch (c)
		{
			case '?':
				usage();
				exit(ExitUsageError);
				break;
			case 'b':
				setBaudRate(optarg);
				break;
			case 'D':
				setDeviceType(optarg);
				break;
			case 'd':
				delimiter_ = optarg;
				break;
			case 'H':
				showReportHeader_ = true;
				break;
			case 'h':
				usage();
				exit(ExitNormal);
				break;
			case 'n':
				deviceNumber_ = atoi(optarg);
				deviceTargeting_ = SingleTarget;

				// FIX THIS, DS; audit against 0?
				break;
			case 'p':
				serialPortName_ = optarg;
				break;
			case 'r':
				setReportType(optarg);
				break;
			case 't':
				timeFormat_ = optarg;
				break;
			case 'v':
				verbose_ = 1;
				break;
			default:
				dieUsage("Unsupported option: %c\n", c);
				break;
		}
	}

	if (optind < argc)
	{
		dieUsage("Unsupported arguments.\n");
	}

	auditArgs();
}

void
Args::auditArgs(void)
{
	if (showReportHeader_ && (delimiter_ == NULL))
	{
		dieUsage("-H requires that the delimiter also be set (-d).\n");
	}

	if ((reportType_ == ReportTypeSystem)
		&& (deviceTargeting_ == SingleTarget))
	{
		dieUsage("The -n option is not allowed with "
			"the system report.\n");
	}
}

void
Args::dieUsage(const char* format, ...)
{
	va_list	ap;
	va_start(ap, format);
	fprintf(stderr, "%s ERROR: ", programName_);
	vfprintf(stderr, format, ap);
	va_end(ap);
	fprintf(stderr, "\n");
	usage();
	exit(ExitUsageError);
}

void
Args::dieMessage(int returnCode, const char* format, ...)
{
	va_list	ap;
	va_start(ap, format);
	fprintf(stderr, "%s ERROR: ", programName_);
	vfprintf(stderr, format, ap);
	va_end(ap);
	exit(returnCode);
}

void Args::setBaudRate(const char* baudRate)
{
	if ((strcmp(baudRate, "2400") == 0)
		|| (strcmp(baudRate, "4800") == 0)
		|| (strcmp(baudRate, "9600") == 0)
		|| (strcmp(baudRate, "19200") == 0))
	{
		baudRate_ = baudRate;
	}
	else
	{
		dieUsage("%s is not a supported baud rate\n.", baudRate);
	}
}

void Args::setReportType(const char* reportType)
{
	if (strcmp(reportType, "now") == 0)
	{
		reportType_ = ReportTypeNow;
	}
	else if (strcmp(reportType, "day") == 0)
	{
		reportType_ = ReportTypeDay;
	}
	else if (strcmp(reportType, "year") == 0)
	{
		reportType_ = ReportTypeYear;
	}
	else if (strcmp(reportType, "total") == 0)
	{
		reportType_ = ReportTypeTotal;
	}
	else if (strcmp(reportType, "all") == 0)
	{
		reportType_ = ReportTypeAll;
	}
	else if (strcmp(reportType, "system") == 0)
	{
		reportType_ = ReportTypeSystem;
	}
	else
	{
		dieUsage("%s is not a supported report type.\n", reportType);
	}
}

void Args::setDeviceType(const char* deviceType)
{
	if (strcmp(deviceType, "inverter") == 0)
	{
		deviceType_ = DeviceTypeInverter;
	}
	else if (strcmp(deviceType, "sensor") == 0)
	{
		deviceType_ = DeviceTypeSensorCard;
	}
	else
	{
		dieUsage("%s is not a supported device type.\n", deviceType);
	}
}

const char*
Args::getBaudRate(void) const
{
	return baudRate_;
}

bool
Args::getVerbose(void) const
{
	return verbose_;
}

Args::ReportType
Args::getReportType(void) const
{
	return reportType_;
}

Args::DeviceType
Args::getDeviceType(void) const
{
	return deviceType_;
}

unsigned char
Args::getDeviceNumber(void) const
{
	assert(deviceTargeting_ == SingleTarget);
	return deviceNumber_;
}

Args::DeviceTargeting
Args::getDeviceTargeting(void) const
{
	return deviceTargeting_;
}
